### PR TITLE
[linux] do not fail when AppArmor policy loading fails (Fixes #1216)

### LINKED ISSUE #1216: Failed to load AppArmor policy
hi everyone,

OS ubuntu 18.04 
❯ uname -a                                                                      ─╯
Linux ..Precision-3520 5.0.21-050021-generic #201906040731 SMP Tue Jun 4 07:33:07 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
❯ multipass version                                                   ─╯
multipass  0.10.0-dev.116+ga77aa05
multipassd 0.10.0-dev.116+ga77aa05
         

multipass worked perfectly until yesterday.

```
multipass launch bionic -n u18cc1                                                                                  ─╯
launch failed: Failed to load AppArmor policy                                   
#include <tunables/global>
profile multipass.dnsmasq flags=(attach_disconnected) {
  #include <abstractions/base>
  #include <abstractions/nameservice>

  capability chown,
  capability net_bind_service,
  capability setgid,
  capability setuid,
  capability dac_override,
  capability dac_read_search,
  capability net_admin,         # for DHCP server
  capability net_raw,           # for DHCP server ping checks
  network inet raw,
  network inet6 raw,

  # Allow multipassd send dnsmasq signals
  signal (receive) peer=snap.multipass.multipassd,

  # access to iface mtu needed for Router Advertisement messages in IPv6
  # Neighbor Discovery protocol (RFC 2461)
  @{PROC}/sys/net/ipv6/conf/*/mtu r,

  # binary and its libs
  /snap/multipass/1434/usr/sbin/dnsmasq ixr,
  /snap/multipass/1434/{usr/,}lib/@{multiarch}/{,**/}*.so* rm,

  # CLASSIC ONLY: need to specify required libs from core snap
  /snap/core18/*/{,usr/}lib/@{multiarch}/{,**/}*.so* rm,

  /var/snap/multipass/common/data/multipassd/network/dnsmasq.leases rw,           # Leases file
  /var/snap/multipass/common/data/multipassd/network/dnsmasq.hosts r,             # Hosts file

  /var/snap/multipass/common/data/multipassd/network/dnsmasq.pid w,     # pid file
}
    : errno=1 ()

❯             
grep apparmor  strace.log                                                     ─╯
newfstatat(AT_FDCWD, "/etc/apparmor.d/cache", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
openat(AT_FDCWD, "/sys/kernel/security/apparmor/features", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/etc/apparmor.d/cache", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
openat(AT_FDCWD, "/sys/kernel/security/apparmor/features", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/local/sbin/apparmor_parser", 0xc420320038, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/local/bin/apparmor_parser", 0xc420320108, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/sbin/apparmor_parser", 0xc4203201d8, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/bin/apparmor_parser", 0xc4203202a8, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/sbin/apparmor_parser", {st_mode=S_IFREG|0755, st_size=1472232, ...}, 0) = 0
newfstatat(AT_FDCWD, "/sbin/apparmor_parser", {st_mode=S_IFREG|0755, st_size=1472232, ...}, 0) = 0
stat("/sys/kernel/security/apparmor", 0x7ffe49af7870) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/sys/module/apparmor/parameters/enabled", O_RDONLY) = 4
newfstatat(AT_FDCWD, "/etc/apparmor.d/cache", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/tls/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/tls/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/tls/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/tls/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/tls/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/tls/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/tls/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/tls/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/tls/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/tls/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/tls/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/tls/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = 3
```

Using libvirt backend multipass works and I can run VM.

I try to re-install , flush snap dir , build from source  without  any success
thanks for you help

### PR DESCRIPTION
**Pull request by Saviq**
---
To test:
```shell
multipass stop --all
sudo modprobe -r kvm_intel
sudo modprobe kvm_intel nested=1
multipass start
multipass shell primary

# inside the instance
wget \
  https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.0.21/linux-headers-5.0.21-050021_5.0.21-050021.201906040731_all.deb \
  https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.0.21/linux-headers-5.0.21-050021-generic_5.0.21-050021.201906040731_amd64.deb \
  https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.0.21/linux-image-unsigned-5.0.21-050021-generic_5.0.21-050021.201906040731_amd64.deb \
  https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.0.21/linux-modules-5.0.21-050021-generic_5.0.21-050021.201906040731_amd64.deb
sudo dpkg -i *.deb
sudo reboot

# enter the instance again
multipass shell primary

# inside the instance, ensure the new kernel is loaded
uname -a

sudo snap install multipass --beta --classic

# this will fail
multipass launch snapcraft:core18 --mem 512M

# refresh to this PR
sudo snap refresh multipass --channel edge/pr1218

# this will now work :)
multipass launch snapcraft:core18 --mem 512M

# you can see the AppArmor warnings (they fired before the launch) in journalctl
journalctl -u snap.multipass*
```

### CODE REVIEW

File: src/platform/backends/shared/linux/process_factory.cpp [hunk 1]
```diff
@@ -90,7 +99,16 @@ std::unique_ptr<mp::Process> mp::ProcessFactory::create_process(std::unique_ptr<
...
     if (apparmor && !process_spec->apparmor_profile().isNull())
     {
-        return std::make_unique<AppArmoredProcess>(apparmor.value(), std::move(process_spec));
+        try
+        {
+            return std::make_unique<AppArmoredProcess>(apparmor.value(), std::move(process_spec));
+        }
+        catch (const mp::AppArmorException& e)
+        {
+            // TODO: This won't fly in strict mode (#1074), since we'll be confined by snapd
```
Comment (ricab) [inline src/platform/backends/shared/linux/process_factory.cpp]: What's the plan then?

File: src/platform/backends/shared/linux/process_factory.cpp [hunk 2]
```diff
@@ -90,7 +99,16 @@ std::unique_ptr<mp::Process> mp::ProcessFactory::create_process(std::unique_ptr<
...
+            mpl::log(mpl::Level::warning, "apparmor", e.what());
+            return std::make_unique<BasicProcess>(std::move(process_spec));
```
Comment (ricab) [inline src/platform/backends/shared/linux/process_factory.cpp]: I am afraid the `process_spec` was already moved, it has a valid but indeterminate state at the this point. Not what we want to create a process with.

File: src/platform/backends/shared/linux/process_factory.cpp (hunk 1 above)
Comment (Saviq) [reply src/platform/backends/shared/linux/process_factory.cpp]: We will have to detect the available AppArmor features and amend the profile to work.

File: src/platform/backends/shared/linux/process_factory.cpp (hunk 2 above)
Comment (Saviq) [reply src/platform/backends/shared/linux/process_factory.cpp]: Ah yes, /me really rushed this and didn't think through…

**Comment by Saviq** [comment]
bors r+

Got some space on the Mac again…

**Pull request merged**